packages/@lwc/integration-karma/test-hydration/mismatches/type-coercion-to-string/x/main/main.html

```diff
@@ -1,4 +1,4 @@
 <template>
     <!-- style attr is validated at compile time -->
-    <p class={num} data-attr={num}>{num}</p>
```

Comment:
Remove the `class` from the hydration test mismatch as before this change the `class={123}` was serialized to  `class="123"` but after this change, the same expression gets serialized to `class=""`. 

As far as I can tell, we are only running the hydration suite for the latest API version. Because of this, there is no way to check the before and after behavior. That said, I don't think this change reduces the coverage of this test as it already checks attribute coercion with the `data-attr`.

Reply:
Why do we want `class={123}` to become nothing, rather than a string? `123` is not an "empty" value, so my baseline expectation is that it would be shown, rather than dropped. Or, if we definitely don't want to allow non-string values, we should throw a TypeError.

Reply:
@wjhsf If possible, I would prefer to avoid making assumptions about what the component author intended.

Reply:
The real question here is whether or not the engine should surface an error when the `class` receives a value that is neither a string, an array nor a JavaScript plain object. I would use "least astonishment principle" here and follow the semantics as defined by other popular UI frameworks and libraries:
- Vue ignores all non-string, array, and objects in both devs and production.
- `classnames` coerce number to string ([commit](https://github.com/JedWatson/classnames/commit/bd1e3440b593e2453dcce60ab9803f6e0f9ab54c)), but ignore other objects in dev and prod.

If we take a step back, I don't expect developers to define class names as numbers. Developers are asking for trouble if they are doing this. For this reason, my recommendation would be to stick to Vue's semantics regarding number handling. 

All of the frameworks and libraries silently ignore all the "invalid" class name values in both dev and prod today and I haven't seen any pushback from the community. It should be safe for LWC to do the same for now. That said, if we see developers struggling with this, we can always add warnings in dev mode and turn this warning into an error in the next major version.

Reply:
This behavior is surprising to me. In the current implementation, numbers are [coerced](https://stackblitz.com/edit/salesforce-lwc-vh2sjx?file=src%2Fmodules%2Fx%2Fapp%2Fapp.html,src%2Fmodules%2Fx%2Fapp%2Fapp.js&title=LWC%20playground) to strings for both `class` and `data-foo`:

```html
<h1 data-foo={num} class={num}>Hello world!</h1>
```

Renders:

```html
<h1 class="123" data-foo="123">Hello world!</h1>
```

Rendering to an empty string definitely seems wrong to me, and it seems wronger to have a hydration mismatch. We have plenty of confusion already with hydration mismatches; IMO we shouldn't be adding on to them (as well as breaking existing functionality unnecessarily). /cc @divmain

(I agree that numbers as `class`es is rare BTW. But I still don't see a reason to break it.)

Reply:
OK wait, I see that `.123` is not a valid CSS selector. So maybe there is no point in rendering `123` as anything but `""`, yeah.

![Screenshot 2024-01-25 at 3 30 31 PM](https://github.com/salesforce/lwc/assets/283842/0dad5271-39f9-4063-876a-a3e631ebf6af)

Reply:
After some thought, I think we should just fix the hydration mismatch issue. A developer may want to have an "invalid" value here, e.g. a value that is either `"foo"` or `false` (e.g. `someCondition && "foo"). The framework should be smart enough to recognize that the server rendered the correct thing if the value is `false`.